File: src/main/java/com/oycm/algorithm/e/enum_right_log_left/basic/Solution_14.scala
package com.oycm.algorithm.e.enum_right_log_left.basic

object Solution_14 {

  /**
   * 3623. 统计梯形的数目 I 1580
   * https://leetcode.cn/problems/count-number-of-trapezoids-i/description/
   *
   * 水平梯形: 至少一对 水平边（平行于 x 轴）
   *
   * @param points 二维整数数组, n = 100000, point(i) = (x, y) 表示第 i 点在 笛卡尔平面上的坐标
   * @return
   */
  def countTrapezoids(points: Array[Array[Int]]): Int = {
    /*
    平行 x 轴，则四点中 y 两两相同，但是又 两两不同
    可以统计相同 y 的点数量 c, 能组成的线段数量 c * (c - 1) / 2
    假设有 3 种 y，记为 y1, y2, y3
    y2 计算 y1 * y2
    y3 = (y1 + y2) * y3

    最多的情况是：两条边 c = n/2，线段数：n ^ 2/8, 答案是 n ^ 4 /64 = 1.5625 * 10 ^ 18 < Long.MaxValue
    */
    val mod = 1000000007
    var ans = 0L
    val map = scala.collection.mutable.Map[Int, Long]()
    for (point <- points) {
      map(point(1)) = map.getOrElse(point(1), 0L) + 1
    }
    var s = 0L
    for ((k, v) <- map) {
      val c = v.toLong * (v - 1) / 2
      ans += c * s
      s += c
    }

    (ans % mod).toInt
  }

  def main(args: Array[String]): Unit = {

  }
}

File: src/main/java/com/oycm/month2026/march/Solution_14.java
package com.oycm.month2026.march;

public class Solution_14 {

    /**
     * 1415. <a href="https://leetcode.cn/problems/the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/description/">长度为 n 的开心字符串中字典序第 k 小的字符串</a> 1576
     * <p>
     * 开心字符串:
     * 仅包含小写字母 ['a', 'b', 'c']
     * 对所有在 1 到 s.length - 1 之间的 i ，满足 s[i] != s[i + 1] （字符串的下标从 1 开始）
     * <p>
     * 将长度为 n 的所有开心字符串按字典序排序，排序后的第 k 个开心字符串，如果长度为 n 的开心字符串少于 k 个，那么请你返回 空字符串
     *
     * @param n
     * @param k
     * @return
     */
    public String getHappyString(int n, int k) {
        /*
        题解
         */
        if (k > 3 << (n - 1)) {
            return "";
        }
        k--;
        char[] ans = new char[n];
        ans[0] = (char) ('a' + (k >> (n - 1)));
        for (int i = 1; i < n; i++) {
            ans[i] = (char) ('a' + (k >> (n - 1 - i) & 1));
            if (ans[i] >= ans[i - 1]) {
                ans[i]++;
            }
        }

        return new String(ans);
    }

}
